File: src/python/_0436_Find_Right_Interval.py
class Solution:
    def findRightInterval(self, intervals: List[List[int]]) -> List[int]:
        starts = sorted((s, i) for i, (s, _) in enumerate(intervals))
        ans = []

        for _, end in intervals:
            idx = bisect_left(starts, (end,))
            ans.append(starts[idx][1] if idx < len(starts) else -1)

        return ans

File: src/_0436_Find_Right_Interval.java
class Solution {
    public int[] findRightInterval(int[][] intervals) {
        int length = intervals.length;
        int[] starts = new int[length];
        Map<Integer, Integer> indices = new HashMap<>();

        for (int i = 0; i < length; i++) {
            starts[i] = intervals[i][0];
            indices.put(intervals[i][0], i);
        }

        Arrays.sort(starts);

        int[] ans = new int[length];

        for (int i = 0; i < length; i++) {
            int idx = Arrays.binarySearch(starts, intervals[i][1]);

            if (idx >= 0)
                ans[i] = indices.get(starts[idx]);
            else if (idx + length >= 0)
                ans[i] = indices.get(starts[-idx - 1]);
            else
                ans[i] = -1;
        }

        return ans;
    }
}
